src/Microsoft.Extensions.DependencyInjection.Ordered/ServiceCollectionOrderedExtensions.cs

```diff
@@ -0,0 +1,194 @@
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            if (implementationType == null)
+            {
+                throw new ArgumentNullException(nameof(implementationType));
+            }
+
+            return AddOrdered(services, (ServiceDescriptor) ServiceDescriptor.Transient(serviceType, implementationType));
+        }
+
+        public static IServiceCollection AddOrdered(
+           this IServiceCollection services,
+           Type serviceType,
+           Func<IServiceProvider, object> implementationFactory)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            if (implementationFactory == null)
+            {
+                throw new ArgumentNullException(nameof(implementationFactory));
+            }
+
+            return AddOrdered(services, (ServiceDescriptor) ServiceDescriptor.Transient(serviceType, implementationFactory));
+        }
+
+        public static IServiceCollection AddOrdered<TService, TImplementation>(
+           this IServiceCollection services,
+           Func<IServiceProvider, TImplementation> implementationFactory)
+           where TService : class
+           where TImplementation : class, TService
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (implementationFactory == null)
+            {
+                throw new ArgumentNullException(nameof(implementationFactory));
+            }
+
+            return AddOrdered(services, (ServiceDescriptor) ServiceDescriptor.Transient(typeof(TService), implementationFactory));
+        }
+
+        public static IServiceCollection AddOrdered(
+            this IServiceCollection collection,
+            ServiceDescriptor descriptor)
+        {
+            if (collection == null)
+            {
+                throw new ArgumentNullException(nameof(collection));
+            }
+
+            if (descriptor == null)
+            {
+                throw new ArgumentNullException(nameof(descriptor));
+            }
+
+            var collectionDescriptor = GetOrderedDescriptor(collection, descriptor.ServiceType);
+            collectionDescriptor.Descriptors.Add(descriptor);
+            return collection;
+        }
+
+        private static OrderedEnumerableServiceDescriptor GetOrderedDescriptor(
+            this IServiceCollection collection,
+            Type serviceType)
+        {
+            var descriptor = (OrderedEnumerableServiceDescriptor)
+                collection.FirstOrDefault(d =>
```

Comment:
`OfType` instead? `collection.OfType<OrderedEnumerableServiceDescriptor>().FirstOrDefault(d => d.ServiceType == serviceType);`. Avoids the cast in the end.